apps/roam/src/components/DiscourseContextOverlay.tsx

```diff
@@ -170,33 +272,106 @@ const DiscourseContextOverlay = ({
+      setHydeFilteredNodes([]);
       return;
     }
     const nodesOnPage = getAllReferencesOnPage(selectedPage);
     const nodes = nodesOnPage
       .map((n) => {
         const node = findDiscourseNode(n.uid);
         if (!node || node.backedBy === "default") return null;
+        if (!validTypes.includes(node.type)) return null;
         return {
           uid: n.uid,
           text: n.text,
           type: node.type,
         };
       })
-      .filter((node) => node !== null)
-      .filter((node) => validTypes.includes(node.type))
-      .filter((node) => !results.some((r) => Object.values(r.results).some((result) => result.uid === node.uid)));
+      .filter((node): node is SuggestedNode => node !== null)
+      .filter(
+        (node) =>
+          !results.some((r) =>
+            Object.values(r.results).some((result) => result.uid === node.uid),
+          ),
+      );
 
-    setSuggestedNodes(nodes);
-  }, [selectedPage, discourseNode, relations]);
+    if (nodes.length > 0 && uniqueRelationTypeTriplets.length > 0) {
+      const performSearch = async () => {
+        setIsSearchingHyde(true);
+        setHydeFilteredNodes([]);
+        try {
+          const foundNodes = await runHydeSearch({
+            currentSuggestions: nodes,
+            currentNodeText: tag,
+            relationDetails: uniqueRelationTypeTriplets,
+          });
+          setHydeFilteredNodes(foundNodes);
+        } catch (error) {
+          console.error(
+            "Error during HyDE search operation in useEffect:",
+            error,
+          );
+          setHydeFilteredNodes([]);
+        } finally {
+          setIsSearchingHyde(false);
+        }
+      };
+      performSearch();
+    }
+  }, [selectedPage, results, validTypes, tag, uniqueRelationTypeTriplets]);
 
-  const handleCreateBlock = async (node: { uid: string; text: string }) => {
+  const handleCreateBlock = async (node: SuggestedNode) => {
     await createBlock({
       parentUid: blockUid,
       node: { text: `[[${node.text}]]` },
     });
-    setSuggestedNodes(suggestedNodes.filter((n) => n.uid !== node.uid));
+    setHydeFilteredNodes(hydeFilteredNodes.filter((n) => n.uid !== node.uid));
+  };
+
+  type RunHydeSearchArgs = {
+    currentSuggestions: SuggestedNode[];
+    currentNodeText: string;
+    relationDetails: RelationDetails[];
+  };
+
+  const runHydeSearch = async ({
+    currentSuggestions,
+    currentNodeText,
+    relationDetails,
+  }: RunHydeSearchArgs): Promise<SuggestedNode[]> => {
+    if (
+      !currentSuggestions.length ||
+      !currentNodeText ||
+      !relationDetails.length
+    ) {
+      return [];
+    }
+
+    try {
+      const candidateNodesForHyde = currentSuggestions.map((node) => {
+        const nodeWithEmbedding = {
+          uid: node.uid,
+          text: node.text,
+          type: node.type,
+          embedding: [] as number[],
+        };
+        return nodeWithEmbedding as CandidateNodeWithEmbedding;
```

Comment:
I know I have previously suggested we use `Result` and work from there.  Generally that is the case.

It looks like in this case, however, that we cannot do that because `embedding` is an array and `Result` does not accept that.

So for this case we could change `CandidateNodeWithEmbedding` to be it's own type which would allow define the variable as this type instead of type asserting.

```typescript
export type CandidateNodeWithEmbedding = {
  uid: string;
  text: string;
  type: string;
  embedding: number[];
};
```

If we continue to use `CandidateNodeWithEmbedding`, we should do this.  (this depends on the previous comment)